internal/api/factory.go

```diff
@@ -0,0 +1,199 @@
+	mappings := mod.Spec.ModuleLoader.Container.KernelMappings
+	foundMapping, err := f.helper.findKernelMapping(mappings, kernelVersion)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find mapping for kernel %s: %v", kernelVersion, err)
+	}
+	mld, err := f.helper.prepareModuleLoaderData(foundMapping, mod, kernelVersion)
+	if err != nil {
+		return nil, fmt.Errorf("failed to prepare module loader data for kernel %s: %v", kernelVersion, err)
+	}
+
+	err = f.helper.replaceTemplates(mld)
+	if err != nil {
+		return nil, fmt.Errorf("failed to replace templates in module loader data for kernel %s: %v", kernelVersion, err)
+	}
+	return mld, nil
+}
+
+type moduleLoaderDataFactoryHelper interface {
+	findKernelMapping(mappings []kmmv1beta1.KernelMapping, kernelVersion string) (*kmmv1beta1.KernelMapping, error)
+	prepareModuleLoaderData(mapping *kmmv1beta1.KernelMapping, mod *kmmv1beta1.Module, kernelVersion string) (*ModuleLoaderData, error)
+	replaceTemplates(mld *ModuleLoaderData) error
+}
+
+type moduleLoaderDataFactoryHelperImpl struct {
+	sectionGetter sectionGetter
+}
+
+func (kh *moduleLoaderDataFactoryHelperImpl) findKernelMapping(mappings []kmmv1beta1.KernelMapping, kernelVersion string) (*kmmv1beta1.KernelMapping, error) {
+	for _, m := range mappings {
+		if m.Literal != "" && m.Literal == kernelVersion {
+			return &m, nil
+		}
+
+		if m.Regexp == "" {
+			continue
+		}
+
+		if matches, err := regexp.MatchString(m.Regexp, kernelVersion); err != nil {
+			return nil, fmt.Errorf("could not match regexp %q against kernel %q: %v", m.Regexp, kernelVersion, err)
+		} else if matches {
+			return &m, nil
+		}
+	}
+
+	return nil, errors.New("no suitable mapping found")
+}
+
+func (kh *moduleLoaderDataFactoryHelperImpl) prepareModuleLoaderData(mapping *kmmv1beta1.KernelMapping, mod *kmmv1beta1.Module, kernelVersion string) (*ModuleLoaderData, error) {
+	var err error
+
+	mld := &ModuleLoaderData{}
+	// prepare the build
+	if mapping.Build != nil || mod.Spec.ModuleLoader.Container.Build != nil {
+		mld.Build = kh.sectionGetter.GetRelevantBuild(mod.Spec.ModuleLoader.Container.Build, mapping.Build)
+	}
+
+	// prepare the sign
+	if mapping.Sign != nil || mod.Spec.ModuleLoader.Container.Sign != nil {
+		mld.Sign, err = kh.sectionGetter.GetRelevantSign(mod.Spec.ModuleLoader.Container.Sign, mapping.Sign, kernelVersion)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get the relevant Sign configuration for kernel %s: %v", kernelVersion, err)
+		}
+	}
+
+	// prepare TLS options
+	mld.RegistryTLS = mapping.RegistryTLS
+	if mapping.RegistryTLS == nil {
+		mld.RegistryTLS = &mod.Spec.ModuleLoader.Container.RegistryTLS
+	}
+
+	//prepare container image
+	mld.ContainerImage = mapping.ContainerImage
+	if mapping.ContainerImage == "" {
+		mld.ContainerImage = mod.Spec.ModuleLoader.Container.ContainerImage
+	}
+
+	mld.KernelVersion = kernelVersion
+	mld.Name = mod.Name
+	mld.Namespace = mod.Namespace
+	mld.ImageRepoSecret = mod.Spec.ImageRepoSecret
```

Comment:
ImagePullPolicy also needs to be set here

Reply:
Fixed.